src/Xamarin.Android.Tools.AndroidSdk/Sdks/AndroidSdkBase.cs

```diff
@@ -139,7 +160,63 @@ protected virtual IEnumerable<string> GetAllAvailableAndroidNdks ()
 			foreach (var sdk in GetAllAvailableAndroidSdks ()) {
 				if (sdk == AndroidSdkPath)
 					continue;
-				yield return Path.Combine (sdk, "ndk-bundle");
+				yield return FindBestNDK (sdk);
+			}
+		}
+
+		string FindBestNDK (string androidSdkPath)
+		{
+			var ndkInstances = new SortedDictionary<Version, string> (Comparer<Version>.Create ((Version l, Version r) => r.CompareTo (l)));
+
+			foreach (string ndkPath in Directory.EnumerateDirectories (androidSdkPath, "ndk*", SearchOption.TopDirectoryOnly)) {
+				if (String.Compare ("ndk-bundle", Path.GetFileName (ndkPath), StringComparison.OrdinalIgnoreCase) == 0) {
+					LoadNDKVersion (ndkPath);
+					continue;
+				}
+
+				if (String.Compare ("ndk", Path.GetFileName (ndkPath), StringComparison.OrdinalIgnoreCase) != 0) {
+					continue;
+				}
+
+				foreach (string versionedNdkPath in Directory.EnumerateDirectories (ndkPath, "*", SearchOption.TopDirectoryOnly)) {
+					LoadNDKVersion (versionedNdkPath);
+				}
+			}
+
+			if (ndkInstances.Count == 0) {
+				return String.Empty;
+			}
+
+			return ndkInstances.First ().Value;
+
+			void LoadNDKVersion (string path)
+			{
+				string propsFilePath = Path.Combine (path, "source.properties");
+				if (!File.Exists (propsFilePath)) {
+					return;
+				}
+
+				foreach (string line in File.ReadLines (propsFilePath)) {
+					string[] parts = line.Split (SourcePropertiesKeyValueSplit, 2, StringSplitOptions.RemoveEmptyEntries);
+					if (parts.Length != 2) {
+						continue;
+					}
+
+					if (String.Compare ("Pkg.Revision", parts[0].Trim (), StringComparison.Ordinal) != 0) {
+						continue;
+					}
+
+					if (!Version.TryParse (parts[1].Trim (), out Version? ndkVer) || ndkVer == null || ndkInstances.ContainsKey (ndkVer)) {
+						continue;
+					}
+
+					if (ndkVer.Major < MinimumCompatibleNDKMajorVersion || ndkVer.Major > MaximumCompatibleNDKMajorVersion) {
+						continue;
+					}
+
+					ndkInstances.Add (ndkVer, path);
```

Comment:
What will this do with duplicates?  It throws `ArgumentException`.

It's possible that to have *both* an `ndk-bundle` dir *and* an `ndk/*` dir for the same version, which will cause this `.Add()` to throw.

Reply:
Look at the surrounding code - we check for duplicates.